snooty/parser.py

```diff
@@ -0,0 +1,362 @@
+import collections
+import docutils.nodes
+import logging
+import multiprocessing
+import os
+import pwd
+import subprocess
+from functools import partial
+from pathlib import Path, PurePath
+from typing import Any, Dict, Tuple, Optional, Set, List, Iterable
+from typing_extensions import Protocol
+import docutils.utils
+
+from . import gizaparser, rstparser, util
+from .gizaparser.nodes import GizaCategory
+from .types import Diagnostic, SerializableType, EmbeddedRstParser, Page, \
+    StaticAsset, ProjectConfigError, ProjectConfig
+
+RST_EXTENSIONS = {'.rst', '.txt'}
+logger = logging.getLogger(__name__)
+
+
+class JSONVisitor:
+    """Node visitor that creates a JSON-serializable structure."""
+    def __init__(self,
+                 project_root: Path,
+                 docpath: PurePath,
+                 document: docutils.nodes.document) -> None:
+        self.project_root = project_root
+        self.docpath = docpath
+        self.document = document
+        self.state: List[Dict[str, Any]] = []
+        self.diagnostics: List[Diagnostic] = []
+        self.static_assets: Set[StaticAsset] = set()
+
+    def dispatch_visit(self, node: docutils.nodes.Node) -> None:
+        node_name = node.__class__.__name__
+        if node_name == 'system_message':
+            level = int(node['level'])
+            if level >= 2:
+                level = Diagnostic.Level.from_docutils(level)
+                msg = node[0].astext()
+                self.diagnostics.append(Diagnostic.create(level, msg, util.get_line(node)))
+            raise docutils.nodes.SkipNode()
+        elif node_name in ('definition', 'field_list'):
+            return
+
+        if node_name == 'document':
+            self.state.append({
+                'type': 'root',
+                'children': [],
+                'position': {
+                    'start': {'line': 0}
+                }
+            })
+            return
+
+        doc: Dict[str, SerializableType] = {
+            'type': node_name,
+            'position': {
+                'start': {'line': util.get_line(node)}
+            }
+        }
+
+        if node_name == 'field':
+            key = node.children[0].astext()
+            value = node.children[1].astext()
+            self.state[-1].setdefault('options', {})[key] = value
+            raise docutils.nodes.SkipNode()
+        elif node_name == 'code':
+            doc['type'] = 'code'
+            doc['lang'] = node['lang']
+            doc['copyable'] = node['copyable']
+            if node['emphasize_lines']:
+                doc['emphasize_lines'] = node['emphasize_lines']
+            doc['value'] = node.astext()
+            self.state[-1]['children'].append(doc)
```

Comment:
In all the other cases you're appending to `self.state`.  Why is this one appending to `state[-1].children` instead?